### ISSUE TITLE
Clarification on I2P and proxy settings in settings_pack

### ISSUE THREAD

**Issue by cas--** [arvidn/libtorrent#1293]
I am attempting to migrate our proxy python code to the new settings pack in 1.1 and have a few questions. It's helpful to get this right to display/hide the correct options to the user. 

 * There is a proxy_type now that can be `i2p_proxy` but there is a separate `i2p_hostname` and `i2p_port`. Does this mean setting `proxy_hostname` and `proxy_port` would have no effect for `i2p`? It would make it easier to code if i2p reused the proxy hostname and port.

 * What would happen if the I2P settings are setup and then also the Proxy settings for a different proxy type?
 * The docs for `force_proxy` mention only `Socks5` and `I2P` for this settings but would this also apply  to `Socks4` and `HTTP` proxies?
   > will only be accepted through a SOCKS5 or I2P proxy

 * I also noticed in C++ code that `force_proxy` will stop `natpmp`, `upnp`, `lsd` and `dht`. Are there any alerts to let us know this has occurred or will we have to assume that these are stopped with a proxy enabled and this setting enabled?

 * The docs for `proxy_hostnames` mention both `Socks5` and `HTTP` but the code only seems to apply to [`Socks5`](https://github.com/arvidn/libtorrent/blob/6751a1eeb1dd79bdc90dbf57cc965a6d1313f82f/src/http_connection.cpp#L551)

 * What about `proxy_peer_connections` and `proxy_tracker_connections`, do they only apply to `Socks5` and not i2p or any other proxy type?

**Comment by arvidn** [comment]
> There is a proxy_type now that can be i2p_proxy but there is a separate i2p_hostname and i2p_port. Does this mean setting proxy_hostname and proxy_port would have no effect for i2p? It would make it easier to code if i2p reused the proxy hostname and port.

I can't recall why those are different. I'm pretty sure no other proxy can be combined with i2p.

> What would happen if the I2P settings are setup and then also the Proxy settings for a different proxy type?

There's only one ``proxy_type`` setting, that's the one determining what to do.

> The docs for force_proxy mention only Socks5 and I2P for this settings but would this also apply to Socks4 and HTTP proxies?
> 
>> will only be accepted through a SOCKS5 or I2P proxy

Those are the only proxy types that support incoming connections.

> I also noticed in C++ code that force_proxy will stop natpmp, upnp, lsd and dht. Are there any alerts to let us know this has occurred or will we have to assume that these are stopped with a proxy enabled and this setting enabled?

I don't believe there are any alerts. The rationale for disabling those is that they inherently are tied to your local machine or local network. the DHT is the odd one out, that probably should not be disabled by this setting.

> The docs for proxy_hostnames mention both Socks5 and HTTP but the code only seems to apply to Socks5

The handling of HTTP hostnames is tightly integrated with the HTTP request itself (it turns into simply passing on the hostname in the request, rather than having a separate lookup step). So it's a bit hard to follow that code, and there's not a lot of shared paths between http and socks5.

> What about proxy_peer_connections and proxy_tracker_connections, do they only apply to Socks5 and not i2p or any other proxy type?

Off hand, i believe they don't apply to i2p, but all other proxies.

**Comment by cas--** [comment]
Thanks that's helped :)

Does `anonymous_mode` and `force_proxy` stay in effect if the proxy_type is `None`?

**Comment by arvidn** [comment]
``anonymous_mode`` definitely is orthogonal to whether a proxy is configured or not. I believe ``force_proxy`` still takes effect without a proxy configured (effectively blocking everything)

**Comment by cas--** [comment]
Thanks, I think that covers me for now. Just need to implement all this information UI-side :D

Can we leave this issue open for now as I might provide a docs pull request just to make it a bit clearer based on the information here.

**Comment by Falcosc** [comment]
@cas-- would be very nice if you could create a pullrequest for the doc change to get the ticked closed. I could try to help you but I'm not sure about my ability to summerice it correctly for the doc.

**Comment by cas--** [comment]
Just going over things again...

#### `force_proxy`
- Closes listen sockets.
- Uses SOCK5 or I2P, if configured.
- Disables UPnP, NAT-PMP & LSD. 
- Disables  DHT

> (if a peer proxy is set up and is run on the same machine as the tracker proxy). 
1. Does that essentially mean it also enables `proxy_tracker_connections` or only if that setting has been enabled?
1. You mentioned that DHT should not be disabled, has this been fixed? 
1. You said there was no alert but what I am looking for is how to indicate to the user that these options have been turned off. It seems currently we just have to make that assumption?
-----
#### `anonymous_mode`
 - `peer_fingerprint` and `user_agent` are not included or set empty.
1. Is this simply a subset of `force_proxy` with the above hidden.
----
#### `proxy_peer_connections`
- Web seeds and BitTorrent peer connections. 
- Does not apply to trackers or DHT traffic.
1. In what instance would you disable this settings, i.e. is it worth exposing to regular end-user?
-----
#### I2P
>>    make it easier to code if i2p reused the proxy hostname and port.

>I can't recall why those are different. I'm pretty sure no other proxy can be combined with i2p.

Should I open an issue ticket for this to be looked into on master branch?

**Comment by cas--** [comment]
There are quite a few references to `tracker proxy` and I just realised there used to be such an option but it was removed, is this what is being referred to?

For example:
> anonymous_mode_alert
>
> This alert is posted when a bittorrent feature is blocked because of the anonymous mode. For instance, if the tracker proxy is not set up, no trackers will be used, because trackers can only be used through proxies when in anonymous mode.

In this particular case is `anonymous_mode_alert` now superfluous?

**Comment by arvidn** [comment]
I believe this alert still applies here, it's just that the tracker will use "the proxy", since there's no separate setting for trackers.

A PR fixing these references would be appreciated! :)

**Comment by cas--** [comment]
> A PR fixing these references would be appreciated! :)

Sure once I've got my rather large Deluge PR merged.

Did you see the questions in the previous comment?
